save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/basic/ProjectSettingsMenu.kt

```diff
@@ -67,43 +71,128 @@ private fun String.toRole() = when (this) {
  * @param deleteProjectCallback
  * @param updateProjectSettings
  * @param openMenuSettingsFlag
- * @param updatePermissions
+ * @param updateErrorMessage
  * @return ReactElement
  */
-@Suppress("TOO_LONG_FUNCTION", "LongMethod", "MAGIC_NUMBER")
+@Suppress(
+    "TOO_LONG_FUNCTION",
+    "LongMethod",
+    "MAGIC_NUMBER",
+    "ComplexMethod"
+)
 fun projectSettingsMenu(
     openMenuSettingsFlag: (isOpen: Boolean) -> Unit,
     deleteProjectCallback: () -> Unit,
     updateProjectSettings: (Project) -> Unit,
-    updatePermissions: (Map<String, Role>) -> Unit,
+    updateErrorMessage: (Response) -> Unit,
 ) = fc<ProjectSettingsMenuProps> { props ->
-    var emailFromInput: String? = props.project.email
-    var isPublic: Boolean = props.project.public
-    var numberOfContainers: String = props.project.numberOfContainers.toString()
     @Suppress("LOCAL_VARIABLE_EARLY_DECLARATION")
-    var permissionsChanged: MutableMap<String, Role> = mutableMapOf()
+    val projectRef = useRef(props.project)
+    val (draftProject, setDraftProject) = useState(props.project)
+    useEffect(arrayOf<dynamic>(props.project)) {
+        if (projectRef.current !== props.project) {
+            setDraftProject(props.project)
+            projectRef.current = props.project
+        }
+    }
 
-    val (users, setUsers) = useState(props.users)
+    val projectPath = props.project.let { "${it.organization.name}/${it.name}" }
+
+    val (changeProjectUsers, setChangeProjectUsers) = useState(false)
+    val (projectUsers, setProjectUsers) = useState(props.projectUsers)
+    val getProjectUsers = useRequest(dependencies = arrayOf(changeProjectUsers)) {
+        val usersFromDb = get(
+            url = "$apiUrl/projects/$projectPath/users",
+            headers = Headers().also {
+                it.set("Accept", "application/json")
+            },
+        )
+            .unsafeMap {
+                it.decodeFromJsonString<List<UserInfo>>()
+            }
+        setProjectUsers(usersFromDb)
+    }
 
-    val getUsers = useRequest {
-        if (props.isOpen != true) {
-            val usersFromDb = get(
-                url = "$apiUrl/projects/${props.project.organization.name}/${props.project.name}/users",
-                headers = Headers().also {
-                    it.set("Accept", "application/json")
-                },
+    val (permissionsChanged, setPermissionsChanged) = useState(mutableMapOf<String, Role>())
```

Comment:
Why mutable map? Mutating state variable without using state setter can lead to confusing behavior, I think.

Reply:
Because I use something like this:
```
setPermissionsChanged { permissionsChanged ->
    permissionsChanged.apply {
        put(userName, target.value.toRole())
    }
}
```
To my mind it seems nice

Reply:
Yeah, but you can't guarantee that the map won't be mutated outside of setter. And it can lead to inconsistency between actual map content and dispalyed data, because mutation of the map won't cause re-rendering. Calling `toMutableMap()` before updating state may be a better option

Reply:
Ok, done